pkg/minikube/update/update.go

```diff
@@ -49,64 +50,86 @@ const (
 	localVersion, err := version.GetSemverVersion()
 	if err != nil {
 		glog.Errorln(err)
-		return
+		return err
 	}
-	MaybeUpdate(output, githubOwner, githubRepo, githubRepo, lastUpdateCheckFilePath, localVersion)
+
+	latestVersion, err := getLatestVersionFromGitHub(githubOwner, githubRepo)
+	if err != nil {
+		return err
+	}
+
+	if checkLatestVersion(localVersion, latestVersion) {
+		return MaybeUpdate(output, githubOwner, githubRepo, latestVersion)
+	}
+	return nil
 }
 
-func MaybeUpdate(output io.Writer, githubOwner, githubRepo, binaryName, lastUpdatePath string, localVersion semver.Version) {
+func checkLatestVersion(localVersion, latestVersion semver.Version) bool {
+	if localVersion.Compare(latestVersion) < 0 {
+		return true
+	} else {
+		return false
+	}
+}
 
-	downloadBinary := binaryName + "-" + runtime.GOOS + "-" + runtime.GOARCH
-	updateLinkPrefix := "https://github.com/" + githubOwner + "/" + githubRepo + "/releases/tag/" + version.VersionPrefix
-	downloadLinkFormat := "https://github.com/" + githubOwner + "/" + githubRepo + "/releases/download/v%s/%s"
+func MaybeUpdate(output io.Writer, githubOwner, githubRepo string, latestVersion semver.Version) error {
+
+	var extName string
 
-	if !shouldCheckURLVersion(lastUpdatePath) {
-		return
+	if runtime.GOOS == "windows" {
+		extName = "zip"
+	} else {
+		extName = "tgz"
 	}
-	latestVersion, err := getLatestVersionFromGitHub(githubOwner, githubRepo)
+
+	archiveName := fmt.Sprintf("minishift-%s-%s-%s.%s", latestVersion, runtime.GOOS, runtime.GOARCH, extName)
+	// updateLinkPrefix := "https://github.com/" + githubOwner + "/" + githubRepo + "/releases/tag/" + version.VersionPrefix
+	downloadLinkFormat := "https://github.com/" + githubOwner + "/" + githubRepo + "/releases/download/v%s/%s"
+
+	fmt.Printf("Updating to version %s\n", latestVersion)
+	// err := updateBinary(latestVersion, archiveName, updateLinkPrefix, downloadLinkFormat)
+
+	// downloadedAchive is a string that has the path to archive on the disk
+	downloadedArchive, dir, err := downloadAndSaveArchive(latestVersion, archiveName, downloadLinkFormat)
 	if err != nil {
-		glog.Errorln(err)
-		return
+		return err
 	}
-	if localVersion.Compare(latestVersion) < 0 {
-		err := writeTimeToFile(lastUpdatePath, time.Now().UTC())
-		if err != nil {
-			fmt.Println("Failed to update last update time")
-		}
-		fmt.Fprintf(output, `A newer version of %s is available. Do you want to
-automatically update from %s%s to %s%s now? [y/N] `,
-			binaryName, version.VersionPrefix, localVersion, version.VersionPrefix, latestVersion)
 
-		var confirm string
-		fmt.Scanln(&confirm)
+	defer os.RemoveAll(dir)
 
-		if strings.ToLower(confirm) == "y" {
-			fmt.Printf("Updating to version %s\n", latestVersion)
-			updateBinary(latestVersion, downloadBinary, updateLinkPrefix, downloadLinkFormat)
-			return
-		}
+	if err := downloadAndVerifyChecksum(downloadedArchive, downloadLinkFormat, archiveName, latestVersion); err != nil {
```

Comment:
This really should be part of _downloadAndSaveArchive_. If the checksum does not match, downloadAndSaveArchive_ should just throw an error. Verifying the checksum is integral part of downloadAndSaveArchive